File: android-app/app/src/main/java/app/calendariociclismo/android/util/SimplifiedGuide.kt
package app.calendariociclismo.android.util

import app.calendariociclismo.android.data.model.ProfileSummit
import app.calendariociclismo.android.data.model.ProfileWaypoint
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.math.abs
import kotlin.math.roundToLong

/** Una fila de la guía simplificada de horarios de paso. */
data class GuideRow(
    val km: Double,
    val kmToGo: Double?,
    // start | climb_foot | summit | intermediate_sprint | bonus_sprint |
    // intermediate_split | cobblestone | sterrato | town | finish
    val type: String,
    val label: String?,
    val category: String?,
    val timeUtc: String?,
    val isEstimated: Boolean,
)

/**
 * Construye la guía simplificada de horarios de paso por los puntos
 * destacados de una jornada. Función pura.
 *
 * Las horas manuales (del rutómetro) viajan en `timeUtc` dentro de cada
 * [ProfileSummit]/[ProfileWaypoint]. Salida y llegada usan los horarios de la
 * jornada. Las horas que no vienen en el rutómetro se ESTIMAN por
 * interpolación lineal por km entre las horas conocidas (anclas).
 *
 * Mantener en PARIDAD con `js/simplified-guide.js` y
 * `ios-app/.../Services/SimplifiedGuide.swift`.
 */
object SimplifiedGuide {

    private val isoOut: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC)

    private fun round1(x: Double): Double = Math.round(x * 10).toDouble() / 10

    private fun parseSeconds(iso: String?): Double? =
        iso?.let { DateFormatting.parseIso(it)?.toEpochMilli()?.toDouble()?.div(1000.0) }

    // Redondea al minuto y serializa a ISO UTC (las horas de rutómetro son HH:mm).
    private fun isoFromSeconds(secs: Double): String {
        val rounded = (secs / 60).roundToLong() * 60
        return isoOut.format(java.time.Instant.ofEpochSecond(rounded))
    }

    // Paridad con `js/perfil-pub.js`.
    private fun isWaypointVisible(type: String, isTimeTrial: Boolean): Boolean = when (type) {
        "kom" -> false
        "intermediate_sprint", "bonus_sprint" -> !isTimeTrial
        "intermediate_split" -> isTimeTrial
        else -> true
    }

    private data class Tmp(
        val km: Double,
        var kmToGo: Double?,
        val type: String,
        val label: String?,
        val category: String?,
        var timeUtc: String?,
        var isEstimated: Boolean,
    )

    fun build(
        distanceKm: Double?,
        neutralStartTimeUtc: String?,
        estimatedFinishTimeUtc: String?,
        summits: List<ProfileSummit>,
        waypoints: List<ProfileWaypoint>,
        primaryType: String?,
    ): List<GuideRow> {
        val isTimeTrial = primaryType == "itt" || primaryType == "ttt"
        val rows = mutableListOf<Tmp>()

        // — Salida (km 0) —
        if (neutralStartTimeUtc != null) {
            rows.add(Tmp(0.0, null, "start", null, null, neutralStartTimeUtc, false))
        }

        // — Puertos: pie (estimado) + cima (manual si la hubiera) —
        for (s in summits) {
            val km = s.km ?: continue
            val startKm = s.startKm
            if (startKm != null && startKm < km) {
                rows.add(Tmp(startKm, null, "climb_foot", s.name, s.category, s.footTimeUtc, s.footTimeUtc == null))
            }
            rows.add(Tmp(km, null, "summit", s.name, s.category, s.timeUtc, s.timeUtc == null))
        }

        // — Waypoints —
        for (w in waypoints) {
            val km = w.km ?: continue
            if (!isWaypointVisible(w.type, isTimeTrial)) continue
            rows.add(Tmp(km, null, w.type, w.name, null, w.timeUtc, w.timeUtc == null))
        }

        // — Llegada (km = distancia) —
        if (estimatedFinishTimeUtc != null && distanceKm != null) {
            rows.add(Tmp(distanceKm, null, "finish", null, null, estimatedFinishTimeUtc, false))
        }

        // — Orden por km (estable: pie < cima por construcción) —
        val sorted = rows.sortedWith(compareBy { it.km })

        // — Deduplicar mismo km y tipo (tol. 0.05 km) —
        val deduped = mutableListOf<Tmp>()
        for (r in sorted) {
            val prev = deduped.lastOrNull()
            if (prev != null && prev.type == r.type && abs(prev.km - r.km) < 0.05) {
                if (prev.timeUtc == null && r.timeUtc != null) deduped[deduped.size - 1] = r
                continue
            }
            deduped.add(r)
        }

        // — kmToGo —
        if (distanceKm != null) {
            for (r in deduped) r.kmToGo = round1(distanceKm - r.km)
        }

        // — Interpolación de horas faltantes (no en CRI/CRE) —
        if (!isTimeTrial) {
            val anchors = deduped.mapNotNull { r ->
                val secs = parseSeconds(r.timeUtc) ?: return@mapNotNull null
                r.km to secs
            }
            if (anchors.size >= 2) {
                for (r in deduped) {
                    if (r.timeUtc != null) continue
                    var prev: Pair<Double, Double>? = null
                    var next: Pair<Double, Double>? = null
                    for (a in anchors) {
                        if (a.first <= r.km && (prev == null || a.first > prev!!.first)) prev = a
                        if (a.first >= r.km && (next == null || a.first < next!!.first)) next = a
                    }
                    val p = prev; val n = next
                    if (p != null && n != null && n.first > p.first) {
                        val t = (r.km - p.first) / (n.first - p.first)
                        r.timeUtc = isoFromSeconds(p.second + t * (n.second - p.second))
                        r.isEstimated = true
                    }
                }
            }
        }

        return deduped.map {
            GuideRow(it.km, it.kmToGo, it.type, it.label, it.category, it.timeUtc, it.isEstimated)
        }
    }

    /** True si la jornada tiene guía que merezca enseñarse. Es **opt-in**:
     *  solo si el editor introdujo al menos UNA hora real del rutómetro en un
     *  punto intermedio (cima/waypoint). Las horas interpoladas no bastan. */
    fun hasGuide(rows: List<GuideRow>): Boolean =
        rows.any { it.type != "start" && it.type != "finish" && it.timeUtc != null && !it.isEstimated }
}

File: ios-app/CalendarioCiclismo/Services/SimplifiedGuide.swift
import Foundation

/// Una fila de la guía simplificada de horarios de paso.
struct GuideRow: Identifiable {
    let km: Double
    let kmToGo: Double?
    /// start | climb_foot | summit | intermediate_sprint | bonus_sprint |
    /// intermediate_split | cobblestone | sterrato | town | finish
    let type: String
    let label: String?
    let category: String?
    let timeUtc: String?
    let isEstimated: Bool

    var id: String { "\(km)-\(type)" }
}

/// Construye la guía simplificada de horarios de paso por los puntos
/// destacados de una jornada. Función pura.
///
/// Las horas manuales (del rutómetro) viajan en `timeUtc` dentro de cada
/// `ProfileSummit`/`ProfileWaypoint`. Salida y llegada usan los horarios de
/// la jornada. Las horas que no vienen en el rutómetro se ESTIMAN por
/// interpolación lineal por km entre las horas conocidas (anclas).
///
/// Mantener en PARIDAD con `js/simplified-guide.js` y
/// `android-app/.../util/SimplifiedGuide.kt`.
enum SimplifiedGuide {

    private struct MutableRow {
        var km: Double
        var kmToGo: Double?
        var type: String
        var label: String?
        var category: String?
        var timeUtc: String?
        var isEstimated: Bool
    }

    private static func round1(_ x: Double) -> Double { (x * 10).rounded() / 10 }

    private static func parseSeconds(_ iso: String?) -> Double? {
        guard let iso else { return nil }
        return DateFormatting.parseISO(iso)?.timeIntervalSince1970
    }

    /// Crea un serializador ISO UTC. Se crea por invocación porque
    /// `ISO8601DateFormatter` (NSObject) no es thread-safe — mismo criterio
    /// que `DateFormatting.makeISOFormatter`.
    private static func makeISOOut() -> ISO8601DateFormatter {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        f.timeZone = TimeZone(identifier: "UTC")
        return f
    }

    // Redondea al minuto y serializa a ISO UTC (las horas de rutómetro son HH:mm).
    private static func isoFromSeconds(_ secs: Double) -> String {
        let rounded = (secs / 60).rounded() * 60
        return makeISOOut().string(from: Date(timeIntervalSince1970: rounded))
    }

    // Paridad con `js/perfil-pub.js`: en CRI/CRE se muestran los puntos
    // intermedios; en el resto, los sprints. Pavé/sterrato/localidad siempre.
    private static func isWaypointVisible(_ type: String, isTimeTrial: Bool) -> Bool {
        switch type {
        case "kom": return false
        case "intermediate_sprint", "bonus_sprint": return !isTimeTrial
        case "intermediate_split": return isTimeTrial
        default: return true
        }
    }

    static func build(
        distanceKm: Double?,
        neutralStartTimeUtc: String?,
        estimatedFinishTimeUtc: String?,
        summits: [ProfileSummit],
        waypoints: [ProfileWaypoint],
        primaryType: String?
    ) -> [GuideRow] {
        let isTimeTrial = primaryType == "itt" || primaryType == "ttt"
        var rows: [MutableRow] = []

        // — Salida (km 0) —
        if let start = neutralStartTimeUtc {
            rows.append(MutableRow(km: 0, kmToGo: nil, type: "start", label: nil,
                                   category: nil, timeUtc: start, isEstimated: false))
        }

        // — Puertos: pie (estimado) + cima (manual si la hubiera) —
        for s in summits {
            guard let km = s.km else { continue }
            if let startKm = s.startKm, startKm < km {
                rows.append(MutableRow(km: startKm, kmToGo: nil, type: "climb_foot",
                                       label: s.name, category: s.category,
                                       timeUtc: s.footTimeUtc, isEstimated: s.footTimeUtc == nil))
            }
            rows.append(MutableRow(km: km, kmToGo: nil, type: "summit", label: s.name,
                                   category: s.category, timeUtc: s.timeUtc,
                                   isEstimated: s.timeUtc == nil))
        }

        // — Waypoints —
        for w in waypoints {
            guard let km = w.km else { continue }
            guard isWaypointVisible(w.type, isTimeTrial: isTimeTrial) else { continue }
            rows.append(MutableRow(km: km, kmToGo: nil, type: w.type, label: w.name,
                                   category: nil, timeUtc: w.timeUtc,
                                   isEstimated: w.timeUtc == nil))
        }

        // — Llegada (km = distancia) —
        if let finish = estimatedFinishTimeUtc, let dist = distanceKm {
            rows.append(MutableRow(km: dist, kmToGo: nil, type: "finish", label: nil,
                                   category: nil, timeUtc: finish, isEstimated: false))
        }

        // — Orden por km (estable: pie < cima por construcción) —
        rows = rows.enumerated()
            .sorted { $0.element.km != $1.element.km ? $0.element.km < $1.element.km : $0.offset < $1.offset }
            .map { $0.element }

        // — Deduplicar mismo km y tipo (tol. 0.05 km) —
        var deduped: [MutableRow] = []
        for r in rows {
            if var prev = deduped.last, prev.type == r.type, abs(prev.km - r.km) < 0.05 {
                if prev.timeUtc == nil, r.timeUtc != nil {
                    prev = r
                    deduped[deduped.count - 1] = prev
                }
                continue
            }
            deduped.append(r)
        }

        // — kmToGo —
        for i in deduped.indices {
            if let dist = distanceKm { deduped[i].kmToGo = round1(dist - deduped[i].km) }
        }

        // — Interpolación de horas faltantes (no en CRI/CRE) —
        if !isTimeTrial {
            let anchors: [(km: Double, secs: Double)] = deduped.compactMap { r in
                guard let secs = parseSeconds(r.timeUtc) else { return nil }
                return (r.km, secs)
            }
            if anchors.count >= 2 {
                for i in deduped.indices where deduped[i].timeUtc == nil {
                    let km = deduped[i].km
                    var prev: (km: Double, secs: Double)?
                    var next: (km: Double, secs: Double)?
                    for a in anchors {
                        if a.km <= km, prev == nil || a.km > prev!.km { prev = a }
                        if a.km >= km, next == nil || a.km < next!.km { next = a }
                    }
                    if let prev, let next, next.km > prev.km {
                        let t = (km - prev.km) / (next.km - prev.km)
                        deduped[i].timeUtc = isoFromSeconds(prev.secs + t * (next.secs - prev.secs))
                        deduped[i].isEstimated = true
                    }
                }
            }
        }

        return deduped.map {
            GuideRow(km: $0.km, kmToGo: $0.kmToGo, type: $0.type, label: $0.label,
                     category: $0.category, timeUtc: $0.timeUtc, isEstimated: $0.isEstimated)
        }
    }

    /// True si la jornada tiene guía que merezca enseñarse. Es **opt-in**:
    /// solo si el editor introdujo al menos UNA hora real del rutómetro en un
    /// punto intermedio (cima/waypoint). Las horas interpoladas no bastan.
    static func hasGuide(_ rows: [GuideRow]) -> Bool {
        rows.contains { $0.type != "start" && $0.type != "finish"
            && $0.timeUtc != nil && !$0.isEstimated }
    }
}
